// fhw-web/lib/customError.js
'use strict';

Object.defineProperty(exports, "__esModule", {
	value: true
});

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

exports.generateErrorPage = generateErrorPage;
exports.isConnectionError = isConnectionError;
exports.NotImplementedError = NotImplementedError;
exports.FileNotFoundError = FileNotFoundError;
exports.FunctionNotFoundError = FunctionNotFoundError;
exports.RessourceNotFoundError = RessourceNotFoundError;
exports.HtmlValidationError = HtmlValidationError;
exports.CssValidationError = CssValidationError;
exports.RouteDefinitionError = RouteDefinitionError;
exports.RouteNotFoundError = RouteNotFoundError;
exports.JsonParseError = JsonParseError;
exports.YamlParseError = YamlParseError;
exports.SessionSaveError = SessionSaveError;
exports.HelperAlreadyDeclared = HelperAlreadyDeclared;
exports.ModuleNotFound = ModuleNotFound;
exports.DataSaveError = DataSaveError;
exports.ControllerReturnValueError = ControllerReturnValueError;
exports.WrongFiletypeError = WrongFiletypeError;

var _helper = require('./helper');

var _htmlEscape = require('html-escape');

var _htmlEscape2 = _interopRequireDefault(_htmlEscape);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var errorColor = '#b30000';
//const htmlEscape = require('html-escape');


function wrapInBody() {
	var innerHtml = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '';

	return '<body style="background-color: ' + errorColor + ';">\n' + innerHtml + '\n</body>';
}

function styleBody(top, tail) {
	return top + ' style="background-color: ' + errorColor + ';" ' + tail;
}

function renderError(html, shouldBeHidden) {
	var style = shouldBeHidden ? 'style="display: none;"' : '';

	return '<!-- Error Description in hidden div below -->\n\n\t\t\t<div title="Error Description" ' + style + '>' + html + '</div>';
}

function generateErrorPage(error) {
	var html = error.html ? error.html : '';

	var rawExtract = error.extract ? error.extract : '';
	var extract = rawExtract.length === 0 ? rawExtract : rawExtract.split('\n').map(function (line, index) {
		var lineNo = index + 1;
		var p = lineNo + ' ' + (0, _htmlEscape2.default)(line);

		return p;
	}).join('\n');

	var stacktrace = error.stack ? error.stack : '';
	stacktrace = (0, _htmlEscape2.default)(stacktrace).split('\n').join('<br>');

	var regex = /(<body)([\s\S]*)/g;
	var match = regex.exec(html);

	if ((0, _helper.isDefined)(match) && (0, _helper.isDefined)(match[1]) && (0, _helper.isDefined)(match[2])) {
		html = styleBody(match[1], match[2]);
	} else {
		html = wrapInBody(html);
	}

	var shouldBeHidden = (0, _helper.isDefined)(error.html) && error.html.length > 0;

	var description = renderError(' \n\t\t<h1>An ' + error.name + ' occured:</h1>\n \n\t\t<code name="stacktrace" style="white-space: pre-line">' + stacktrace + '</code>\n\n\t\t<pre name="extract">' + extract + '</pre>\n\n\t\t<div name="raw-extract" style="display: none;">' + (0, _htmlEscape2.default)(rawExtract) + '</div>\n\t', shouldBeHidden);

	return '' + description + html;
}

// based on https://stackoverflow.com/a/32749533

var ExtendableError = function (_Error) {
	(0, _inherits3.default)(ExtendableError, _Error);

	function ExtendableError(message) {
		var status = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : 500;
		var html = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : '';
		var extract = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : '';
		(0, _classCallCheck3.default)(this, ExtendableError);

		var _this = (0, _possibleConstructorReturn3.default)(this, (ExtendableError.__proto__ || (0, _getPrototypeOf2.default)(ExtendableError)).call(this, message));

		_this.name = _this.constructor.name;
		_this.status = status;
		_this.html = html;
		_this.extract = extract;

		console.log(_this.name + ': ' + message);

		if (typeof Error.captureStackTrace === 'function') {
			Error.captureStackTrace(_this, _this.constructor);
		} else {
			_this.stack = new Error(message).stack;
		}
		return _this;
	}

	return ExtendableError;
}(Error);

var NotImplementedError = function (_ExtendableError) {
	(0, _inherits3.default)(NotImplementedError, _ExtendableError);

	function NotImplementedError() {
		(0, _classCallCheck3.default)(this, NotImplementedError);
		return (0, _possibleConstructorReturn3.default)(this, (NotImplementedError.__proto__ || (0, _getPrototypeOf2.default)(NotImplementedError)).apply(this, arguments));
	}

	return NotImplementedError;
}(ExtendableError);

var RessourceNotFoundError = function (_ExtendableError2) {
	(0, _inherits3.default)(RessourceNotFoundError, _ExtendableError2);

	function RessourceNotFoundError() {
		(0, _classCallCheck3.default)(this, RessourceNotFoundError);
		return (0, _possibleConstructorReturn3.default)(this, (RessourceNotFoundError.__proto__ || (0, _getPrototypeOf2.default)(RessourceNotFoundError)).apply(this, arguments));
	}

	return RessourceNotFoundError;
}(ExtendableError);

var FileNotFoundError = function (_ExtendableError3) {
	(0, _inherits3.default)(FileNotFoundError, _ExtendableError3);

	function FileNotFoundError() {
		(0, _classCallCheck3.default)(this, FileNotFoundError);
		return (0, _possibleConstructorReturn3.default)(this, (FileNotFoundError.__proto__ || (0, _getPrototypeOf2.default)(FileNotFoundError)).apply(this, arguments));
	}

	return FileNotFoundError;
}(ExtendableError);

var FunctionNotFoundError = function (_ExtendableError4) {
	(0, _inherits3.default)(FunctionNotFoundError, _ExtendableError4);

	function FunctionNotFoundError() {
		(0, _classCallCheck3.default)(this, FunctionNotFoundError);
		return (0, _possibleConstructorReturn3.default)(this, (FunctionNotFoundError.__proto__ || (0, _getPrototypeOf2.default)(FunctionNotFoundError)).apply(this, arguments));
	}

	return FunctionNotFoundError;
}(ExtendableError);

var HtmlValidationError = function (_ExtendableError5) {
	(0, _inherits3.default)(HtmlValidationError, _ExtendableError5);

	function HtmlValidationError() {
		(0, _classCallCheck3.default)(this, HtmlValidationError);
		return (0, _possibleConstructorReturn3.default)(this, (HtmlValidationError.__proto__ || (0, _getPrototypeOf2.default)(HtmlValidationError)).apply(this, arguments));
	}

	return HtmlValidationError;
}(ExtendableError);

var CssValidationError = function (_ExtendableError6) {
	(0, _inherits3.default)(CssValidationError, _ExtendableError6);

	function CssValidationError() {
		(0, _classCallCheck3.default)(this, CssValidationError);
		return (0, _possibleConstructorReturn3.default)(this, (CssValidationError.__proto__ || (0, _getPrototypeOf2.default)(CssValidationError)).apply(this, arguments));
	}

	return CssValidationError;
}(ExtendableError);

var RouteDefinitionError = function (_ExtendableError7) {
	(0, _inherits3.default)(RouteDefinitionError, _ExtendableError7);

	function RouteDefinitionError() {
		(0, _classCallCheck3.default)(this, RouteDefinitionError);
		return (0, _possibleConstructorReturn3.default)(this, (RouteDefinitionError.__proto__ || (0, _getPrototypeOf2.default)(RouteDefinitionError)).apply(this, arguments));
	}

	return RouteDefinitionError;
}(ExtendableError);

var RouteNotFoundError = function (_ExtendableError8) {
	(0, _inherits3.default)(RouteNotFoundError, _ExtendableError8);

	function RouteNotFoundError() {
		(0, _classCallCheck3.default)(this, RouteNotFoundError);
		return (0, _possibleConstructorReturn3.default)(this, (RouteNotFoundError.__proto__ || (0, _getPrototypeOf2.default)(RouteNotFoundError)).apply(this, arguments));
	}

	return RouteNotFoundError;
}(ExtendableError);

var JsonParseError = function (_ExtendableError9) {
	(0, _inherits3.default)(JsonParseError, _ExtendableError9);

	function JsonParseError() {
		(0, _classCallCheck3.default)(this, JsonParseError);
		return (0, _possibleConstructorReturn3.default)(this, (JsonParseError.__proto__ || (0, _getPrototypeOf2.default)(JsonParseError)).apply(this, arguments));
	}

	return JsonParseError;
}(ExtendableError);

var YamlParseError = function (_ExtendableError10) {
	(0, _inherits3.default)(YamlParseError, _ExtendableError10);

	function YamlParseError() {
		(0, _classCallCheck3.default)(this, YamlParseError);
		return (0, _possibleConstructorReturn3.default)(this, (YamlParseError.__proto__ || (0, _getPrototypeOf2.default)(YamlParseError)).apply(this, arguments));
	}

	return YamlParseError;
}(ExtendableError);

var SessionSaveError = function (_ExtendableError11) {
	(0, _inherits3.default)(SessionSaveError, _ExtendableError11);

	function SessionSaveError() {
		(0, _classCallCheck3.default)(this, SessionSaveError);
		return (0, _possibleConstructorReturn3.default)(this, (SessionSaveError.__proto__ || (0, _getPrototypeOf2.default)(SessionSaveError)).apply(this, arguments));
	}

	return SessionSaveError;
}(ExtendableError);

var HelperAlreadyDeclared = function (_ExtendableError12) {
	(0, _inherits3.default)(HelperAlreadyDeclared, _ExtendableError12);

	function HelperAlreadyDeclared() {
		(0, _classCallCheck3.default)(this, HelperAlreadyDeclared);
		return (0, _possibleConstructorReturn3.default)(this, (HelperAlreadyDeclared.__proto__ || (0, _getPrototypeOf2.default)(HelperAlreadyDeclared)).apply(this, arguments));
	}

	return HelperAlreadyDeclared;
}(ExtendableError);

var ModuleNotFound = function (_ExtendableError13) {
	(0, _inherits3.default)(ModuleNotFound, _ExtendableError13);

	function ModuleNotFound() {
		(0, _classCallCheck3.default)(this, ModuleNotFound);
		return (0, _possibleConstructorReturn3.default)(this, (ModuleNotFound.__proto__ || (0, _getPrototypeOf2.default)(ModuleNotFound)).apply(this, arguments));
	}

	return ModuleNotFound;
}(ExtendableError);

var DataSaveError = function (_ExtendableError14) {
	(0, _inherits3.default)(DataSaveError, _ExtendableError14);

	function DataSaveError() {
		(0, _classCallCheck3.default)(this, DataSaveError);
		return (0, _possibleConstructorReturn3.default)(this, (DataSaveError.__proto__ || (0, _getPrototypeOf2.default)(DataSaveError)).apply(this, arguments));
	}

	return DataSaveError;
}(ExtendableError);

var ControllerReturnValueError = function (_ExtendableError15) {
	(0, _inherits3.default)(ControllerReturnValueError, _ExtendableError15);

	function ControllerReturnValueError() {
		(0, _classCallCheck3.default)(this, ControllerReturnValueError);
		return (0, _possibleConstructorReturn3.default)(this, (ControllerReturnValueError.__proto__ || (0, _getPrototypeOf2.default)(ControllerReturnValueError)).apply(this, arguments));
	}

	return ControllerReturnValueError;
}(ExtendableError);

var WrongFiletypeError = function (_ExtendableError16) {
	(0, _inherits3.default)(WrongFiletypeError, _ExtendableError16);

	function WrongFiletypeError() {
		(0, _classCallCheck3.default)(this, WrongFiletypeError);
		return (0, _possibleConstructorReturn3.default)(this, (WrongFiletypeError.__proto__ || (0, _getPrototypeOf2.default)(WrongFiletypeError)).apply(this, arguments));
	}

	return WrongFiletypeError;
}(ExtendableError);

function isConnectionError(error) {
	return error.code === 'ENOTFOUND' && error.syscall === 'getaddrinfo';
}

function NotImplementedError(message) {
	return new NotImplementedError(message, 500);
}
function FileNotFoundError(message) {
	return new FileNotFoundError(message, 404);
}
function FunctionNotFoundError(message) {
	return new FunctionNotFoundError(message, 500);
}
function RessourceNotFoundError(message) {
	return new RessourceNotFoundError(message, 404);
}
function HtmlValidationError(message, html) {
	var extract = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : html;
	return new HtmlValidationError(message, 500, html, extract);
}
function CssValidationError(message, html, extract) {
	return new CssValidationError(message, 500, html, extract);
}
function RouteDefinitionError(message) {
	return new RouteDefinitionError(message, 500);
}
function RouteNotFoundError(message) {
	return new RouteNotFoundError(message, 500);
}
function JsonParseError(filename, message) {
	return new JsonParseError('JsonParseError in file ' + filename + ': ' + message, 500);
}
function YamlParseError(filename, message) {
	return new YamlParseError('YamlParseError in file ' + filename + ': ' + message, 500);
}
function SessionSaveError(message) {
	return new SessionSaveError(message, 500);
}
function HelperAlreadyDeclared(message) {
	return new HelperAlreadyDeclared(message, 500);
}
function ModuleNotFound(message) {
	return new ModuleNotFound(message, 500);
}
function DataSaveError(filename, data, message) {
	return new DataSaveError('DataSaveError: could not save ' + filename + ' with data ' + data + ': ' + message, 500);
}
function ControllerReturnValueError(message) {
	return new ControllerReturnValueError(message, 500);
}
function WrongFiletypeError(message) {
	return new WrongFiletypeError(message, 500);
}
